contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,325 @@
+    uint256 private amountOutCached;
+
+    // v4 Singleton contract
+    IPoolManager public immutable manager;
+
+    /// @dev custom error function selector length
+    uint256 internal constant MINIMUM_CUSTOM_ERROR_LENGTH = 4;
+
+    /// @dev function selector + length of bytes as uint256 + min length of revert reason padded to multiple of 32 bytes
+    uint256 internal constant MINIMUM_REASON_LENGTH = 68;
+
+    /// @dev min valid reason is 3-words long
+    /// @dev int128[2] + sqrtPriceX96After padded to 32bytes + intializeTicksLoaded padded to 32bytes
+    uint256 internal constant MINIMUM_VALID_RESPONSE_LENGTH = 96;
+
+    /// @dev Only this address may call this function
+    modifier selfOnly() {
+        if (msg.sender != address(this)) revert NotSelf();
+        _;
+    }
+
+    constructor(address _poolManager) {
+        manager = IPoolManager(_poolManager);
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactInputSingle(QuoteExactSingleParams memory params)
+        public
+        override
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded)
+    {
+        try manager.lock(address(this), abi.encodeWithSelector(this._quoteExactInputSingle.selector, params)) {}
+        catch (bytes memory reason) {
+            return _handleRevertSingle(reason);
+        }
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactInput(QuoteExactParams memory params)
+        external
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        try manager.lock(address(this), abi.encodeWithSelector(this._quoteExactInput.selector, params)) {}
+        catch (bytes memory reason) {
+            return _handleRevert(reason);
+        }
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactOutputSingle(QuoteExactSingleParams memory params)
+        public
+        override
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded)
+    {
+        if (params.sqrtPriceLimitX96 == 0) amountOutCached = params.exactAmount;
+
+        try manager.lock(address(this), abi.encodeWithSelector(this._quoteExactOutputSingle.selector, params)) {}
+        catch (bytes memory reason) {
+            if (params.sqrtPriceLimitX96 == 0) delete amountOutCached;
+            return _handleRevertSingle(reason);
+        }
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactOutput(QuoteExactParams memory params)
+        public
+        override
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        try manager.lock(address(this), abi.encodeWithSelector(this._quoteExactOutput.selector, params)) {}
```

Comment:
why doesnt this one need amountOutCached?

Reply:
ohh true, moved amountOutCached into the inner call (_quote...) so quoteExactOutput also uses it